File: wrappers/swift/VirgilCrypto/VirgilCryptoFoundation/KeyAlgFactory.swift
/// Copyright (C) 2015-2019 Virgil Security, Inc.
///
/// All rights reserved.
///
/// Redistribution and use in source and binary forms, with or without
/// modification, are permitted provided that the following conditions are
/// met:
///
///     (1) Redistributions of source code must retain the above copyright
///     notice, this list of conditions and the following disclaimer.
///
///     (2) Redistributions in binary form must reproduce the above copyright
///     notice, this list of conditions and the following disclaimer in
///     the documentation and/or other materials provided with the
///     distribution.
///
///     (3) Neither the name of the copyright holder nor the names of its
///     contributors may be used to endorse or promote products derived from
///     this software without specific prior written permission.
///
/// THIS SOFTWARE IS PROVIDED BY THE AUTHOR ''AS IS'' AND ANY EXPRESS OR
/// IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
/// WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
/// DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
/// INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
/// (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
/// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
/// HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
/// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
/// IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
/// POSSIBILITY OF SUCH DAMAGE.
///
/// Lead Maintainer: Virgil Security Inc. <support@virgilsecurity.com>


import Foundation
import VSCFoundation

/// Create a bridge between "raw keys" and algorithms that can import them.
@objc(VSCFKeyAlgFactory) public class KeyAlgFactory: NSObject {

    /// Create a key algorithm based on an identifier.
    @objc public static func createFromAlgId(algId: AlgId, random: Random) throws -> KeyAlg {
        var error: vscf_error_t = vscf_error_t()
        vscf_error_reset(&error)

        let proxyResult = vscf_key_alg_factory_create_from_alg_id(vscf_alg_id_t(rawValue: UInt32(algId.rawValue)), random.c_ctx, &error)

        try FoundationError.handleStatus(fromC: error.status)

        return FoundationImplementation.wrapKeyAlg(take: proxyResult!)
    }

    /// Create a key algorithm correspond to a specific key.
    @objc public static func createFromKey(key: Key, random: Random) throws -> KeyAlg {
        var error: vscf_error_t = vscf_error_t()
        vscf_error_reset(&error)

        let proxyResult = vscf_key_alg_factory_create_from_key(key.c_ctx, random.c_ctx, &error)

        try FoundationError.handleStatus(fromC: error.status)

        return FoundationImplementation.wrapKeyAlg(take: proxyResult!)
    }

    /// Create a key algorithm that can import "raw public key".
    @objc public static func createFromRawPublicKey(publicKey: RawPublicKey, random: Random) throws -> KeyAlg {
        var error: vscf_error_t = vscf_error_t()
        vscf_error_reset(&error)

        let proxyResult = vscf_key_alg_factory_create_from_raw_public_key(publicKey.c_ctx, random.c_ctx, &error)

        try FoundationError.handleStatus(fromC: error.status)

        return FoundationImplementation.wrapKeyAlg(take: proxyResult!)
    }

    /// Create a key algorithm that can import "raw private key".
    @objc public static func createFromRawPrivateKey(privateKey: RawPrivateKey, random: Random) throws -> KeyAlg {
        var error: vscf_error_t = vscf_error_t()
        vscf_error_reset(&error)

        let proxyResult = vscf_key_alg_factory_create_from_raw_private_key(privateKey.c_ctx, random.c_ctx, &error)

        try FoundationError.handleStatus(fromC: error.status)

        return FoundationImplementation.wrapKeyAlg(take: proxyResult!)
    }
}

File: wrappers/java/foundation/src/main/java/com/virgilsecurity/crypto/foundation/KeyAlgFactory.java
/*
* Copyright (C) 2015-2020 Virgil Security, Inc.
*
* All rights reserved.
*
* Redistribution and use in source and binary forms, with or without
* modification, are permitted provided that the following conditions are
* met:
*
* (1) Redistributions of source code must retain the above copyright
* notice, this list of conditions and the following disclaimer.
*
* (2) Redistributions in binary form must reproduce the above copyright
* notice, this list of conditions and the following disclaimer in
* the documentation and/or other materials provided with the
* distribution.
*
* (3) Neither the name of the copyright holder nor the names of its
* contributors may be used to endorse or promote products derived from
* this software without specific prior written permission.
*
* THIS SOFTWARE IS PROVIDED BY THE AUTHOR ''AS IS'' AND ANY EXPRESS OR
* IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
* WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
* DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
* INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
* (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
* SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
* HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
* STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
* IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
* POSSIBILITY OF SUCH DAMAGE.
*
* Lead Maintainer: Virgil Security Inc. <support@virgilsecurity.com>
*/

package com.virgilsecurity.crypto.foundation;

/*
* Create a bridge between "raw keys" and algorithms that can import them.
*/
public class KeyAlgFactory {

    /*
    * Create a key algorithm based on an identifier.
    */
    public static KeyAlg createFromAlgId(AlgId algId, Random random) throws FoundationException {
        return FoundationJNI.INSTANCE.keyAlgFactory_createFromAlgId(algId, random);
    }

    /*
    * Create a key algorithm correspond to a specific key.
    */
    public static KeyAlg createFromKey(Key key, Random random) throws FoundationException {
        return FoundationJNI.INSTANCE.keyAlgFactory_createFromKey(key, random);
    }

    /*
    * Create a key algorithm that can import "raw public key".
    */
    public static KeyAlg createFromRawPublicKey(RawPublicKey publicKey, Random random) throws FoundationException {
        return FoundationJNI.INSTANCE.keyAlgFactory_createFromRawPublicKey(publicKey, random);
    }

    /*
    * Create a key algorithm that can import "raw private key".
    */
    public static KeyAlg createFromRawPrivateKey(RawPrivateKey privateKey, Random random) throws FoundationException {
        return FoundationJNI.INSTANCE.keyAlgFactory_createFromRawPrivateKey(privateKey, random);
    }
}

